### PR TITLE
Add Tinkerbell label selector and configure disk type in the template…

### PR DESCRIPTION
**Pull request by panktishah26**
*Description of changes:*
In this PR, I have added Tinkerbell label selector in the cluster config file and configured disk type in the template config field.

### CODE REVIEW

File: pkg/api/v1alpha1/cluster_types.go [hunk 1]
```diff
@@ -613,6 +613,10 @@ type Ref struct {
 	Name string `json:"name,omitempty"`
 }
 
+type HardwareSelector struct {
```
Comment (chrisdoherty4) [inline pkg/api/v1alpha1/cluster_types.go:616]: Selectors need to be `map[string]string` to allow arbitrary key-value pairs.

**Review by chrisdoherty4 — CHANGES_REQUESTED** [review]
Principally, this is on the right track - just needs a few adjustments.

We need selectors to be completely definable by the user. The user should control, entirely, what key-value pairs are used in Hardware CSV labels (mainline already has this), _and_ what key-value pairs are used in selectors to match those labels.

Having done that, for each selector we cache the first observed Disk from a matching Machine instance. An example of a selector-label matching algorithm can be found in the [PR I just raised](https://github.com/aws/eks-anywhere/pull/2299/files#diff-05d6aded64a4a29b5fd8c21346546229256a812aa21b9656d1ef238c7ad2865dR290) and we should probably use the same function.

Having cached the disks we can pass a construct with access to that cache to the template builder and _provide lookup behavior_. It is extremely important that we provide _behavior_ (methods) and don't wangle inner fields because we start to violate the [Law of Demeter](https://en.wikipedia.org/wiki/Law_of_Demeter).

File: pkg/api/v1alpha1/cluster_types.go [hunk 2]
```diff
@@ -613,6 +613,8 @@ type Ref struct {
 	Name string `json:"name,omitempty"`
 }
 
+type HardwareSelector map[string]string
+
```
Comment (abhinavmpandey08) [inline pkg/api/v1alpha1/cluster_types.go:617]: Can you define this type in `tinkerbellmachineconfig_types.go` instead?

File: pkg/api/v1alpha1/tinkerbelltemplateconfig_defaults.go [hunk 3]
```diff
@@ -29,39 +31,39 @@ warnings:
...
-func withNetplanAction(b v1alpha1.VersionsBundle) ActionOpt {
+func withNetplanAction(b v1alpha1.VersionsBundle, diskType string) ActionOpt {
 	return func(a *[]tinkerbell.Action) {
 		*a = append(*a, tinkerbell.Action{
 			Name:    "write-netplan",
 			Image:   b.Tinkerbell.TinkerbellStack.Actions.WriteFile.URI,
 			Timeout: 90,
 			Environment: map[string]string{
-				"DEST_DISK": "/dev/sda2",
+				"DEST_DISK": fmt.Sprintf("%s2", diskType),
```
Comment (abhinavmpandey08) [inline pkg/api/v1alpha1/tinkerbelltemplateconfig_defaults.go:66]: We are just appending `2` to `diskType`. Do we account for whether `diskType` is `sda` or `nvme`?
Because for `sda` you can just append 2 but with nvme you need `p2`

File: pkg/api/v1alpha1/tinkerbelltemplateconfig_defaults_test.go [hunk 4]
```diff
@@ -10,6 +10,7 @@ import (
 
 func TestWithDefaultActionsFromBundle(t *testing.T) {
 	vBundle := givenVersionBundle()
+	diskType := "/dev/sda"
```
Comment (abhinavmpandey08) [inline pkg/api/v1alpha1/tinkerbelltemplateconfig_defaults_test.go:13]: Should add a test case with nvme as well

File: pkg/providers/tinkerbell/config/template-cp.yaml [hunk 5]
```diff
@@ -178,6 +178,12 @@ spec:
   template:
     {{- if and .etcdTemplateOverride (ne .format "")}}
     spec:
+      hardwareAffinity:
+        required:
+        - labelSelector:
+            matchLabels: {{ range $key, $value := .etcdHardwareSelector}}
+              {{ $key }}: {{ $value}}
+            {{- end }}
```
Comment (abhinavmpandey08) [inline pkg/providers/tinkerbell/config/template-cp.yaml:186]: Have you checked what happens here if the `etcdHardwareSelector` is empty?

File: pkg/api/v1alpha1/tinkerbelltemplateconfig_defaults.go (hunk 3 above)
Comment (chrisdoherty4) [reply pkg/api/v1alpha1/tinkerbelltemplateconfig_defaults.go:66]: We talked about dealing with NVMe on a separate PR so we can unblock mainline.

File: pkg/api/v1alpha1/tinkerbelltemplateconfig_defaults_test.go (hunk 4 above)
Comment (chrisdoherty4) [reply pkg/api/v1alpha1/tinkerbelltemplateconfig_defaults_test.go:13]: https://github.com/aws/eks-anywhere/pull/2300#discussion_r889709080

File: pkg/api/v1alpha1/tinkerbellmachineconfig.go [hunk 6]
```diff
@@ -87,3 +87,9 @@ func WithTemplateRef(ref ProviderRefAccessor) TinkerbellMachineConfigGenerateOpt
 		}
 	}
 }
+
+func WithHardwareLabel() TinkerbellMachineConfigGenerateOpt {
```
Comment (chrisdoherty4) [inline pkg/api/v1alpha1/tinkerbellmachineconfig.go:91]: Map initialization shoudln't be optional. Could we move this to the `New*`.

File: pkg/api/v1alpha1/tinkerbelltemplateconfig.go [hunk 7]
```diff
@@ -18,7 +18,7 @@ const TinkerbellTemplateConfigKind = "TinkerbellTemplateConfig"
 type ActionOpt func(action *[]tinkerbell.Action)
 
 // NewDefaultTinkerbellTemplateConfigCreate returns a default TinkerbellTemplateConfig with the required Tasks and Actions
-func NewDefaultTinkerbellTemplateConfigCreate(name string, versionBundle v1alpha1.VersionsBundle) *TinkerbellTemplateConfig {
+func NewDefaultTinkerbellTemplateConfigCreate(name string, versionBundle v1alpha1.VersionsBundle, diskType string) *TinkerbellTemplateConfig {
```
Comment (chrisdoherty4) [inline pkg/api/v1alpha1/tinkerbelltemplateconfig.go:21]: Could we change to using `disk` or `diskPath`. We aren't really describing a type of disk. Also a code comment everywhere this is accepted, given its just a string, would go a long way so we establish expectations.

File: pkg/api/v1alpha1/tinkerbelltemplateconfig_defaults.go [hunk 8]
```diff
@@ -74,14 +76,14 @@ func withNetplanAction(b v1alpha1.VersionsBundle) ActionOpt {
...
-func withTinkCloudInitAction(b v1alpha1.VersionsBundle) ActionOpt {
+func withTinkCloudInitAction(b v1alpha1.VersionsBundle, diskType string) ActionOpt {
 	return func(a *[]tinkerbell.Action) {
 		*a = append(*a, tinkerbell.Action{
 			Name:    "add-tink-cloud-init-config",
 			Image:   b.Tinkerbell.TinkerbellStack.Actions.WriteFile.URI,
 			Timeout: 90,
 			Environment: map[string]string{
-				"DEST_DISK": "/dev/sda2",
+				"DEST_DISK": fmt.Sprintf("%s2", diskType),
```
Comment (chrisdoherty4) [inline pkg/api/v1alpha1/tinkerbelltemplateconfig_defaults.go:86]: The partition number shouldn't be the responsibility of individual actions to define. They should just accept a destination disk, in full. 

The partition number, for now, could be defined by the templating construct perhaps? We could modify this in a subsequent PR as we explore supporting multiple disk types (NVMe etc).

File: pkg/providers/tinkerbell/hardware/catalogue_hardware.go [hunk 9]
```diff
@@ -1,21 +1,67 @@
...
+
+	tinkv1alpha1 "github.com/tinkerbell/tink/pkg/apis/core/v1alpha1"
 	corev1 "k8s.io/api/core/v1"
 	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 
+	eksav1alpha1 "github.com/aws/eks-anywhere/pkg/api/v1alpha1"
 	"github.com/aws/eks-anywhere/pkg/constants"
 )
 
+type DiskExtractor struct {
```
Comment (chrisdoherty4) [inline pkg/providers/tinkerbell/hardware/catalogue_hardware.go:14]: Can we get code documentation on _all_ exported types at a minimum. The docs should explain what the scope of responsibility is and any behavior expectations.

File: pkg/providers/tinkerbell/create.go [hunk 10]
```diff
@@ -118,7 +118,14 @@ func (p *Provider) SetupAndValidateCreateCluster(ctx context.Context, clusterSpe
 		validator.Register(NewIPNotInUseAssertion(p.netClient))
 	}
 
-	writer := hardware.NewMachineCatalogueWriter(p.catalogue)
+	catalogueWriter := hardware.NewMachineCatalogueWriter(p.catalogue)
+
+	for _, mc := range p.machineConfigs {
+		p.templateBuilder.diskExtractor.RegisterHardwareSelector(mc.Spec.HardwareSelector)
```
Comment (chrisdoherty4) [inline pkg/providers/tinkerbell/create.go:124]: We shouldn't be reaching into other constructs and modifying their members (Law of Demeter). 

Its not a great solution, but for now can we expose a `RegisterHardwareSelector` on the `TemplateBuilder` so we can call `p.templateBuilder.RegisterHardwareSelector()`. This, at least, maintains a degree of separation that will compliment any `TemplateBuilder` dependency injection we refactor for in the future.

File: pkg/providers/tinkerbell/hardware/catalogue_hardware.go (hunk 9 above)
Comment (chrisdoherty4) [inline pkg/providers/tinkerbell/hardware/catalogue_hardware.go:14]: The `DiskExtractor` needs a suite of unit tests, positive and negative.

File: pkg/providers/tinkerbell/hardware/catalogue_hardware.go [hunk 11]
```diff
@@ -1,21 +1,67 @@
...
+	for key, val := range hardwareSelector {
+		selectorKey[key] = val
+	}
+	return selectorKey.String()
+}
+
+func (d *DiskExtractor) GetDisk(hardwareSelector eksav1alpha1.HardwareSelector) (string, error) {
+	key := serializedHardwareSelector(hardwareSelector)
+	if _, ok := d.Disks[key]; !ok {
+		return "", fmt.Errorf("hardware selector does not match with any given hardware machines %v", hardwareSelector)
```
Comment (chrisdoherty4) [inline pkg/providers/tinkerbell/hardware/catalogue_hardware.go:52]: Use colon's in errors to separate added context from original error.

```suggestion
		return "", fmt.Errorf("hardware selector does not match with any given hardware machines: %v", hardwareSelector)
```

File: pkg/providers/tinkerbell/hardware/catalogue_hardware.go [hunk 12]
```diff
@@ -1,21 +1,67 @@
...
+	}
+	return d.Disks[key], nil
```
Comment (chrisdoherty4) [inline pkg/providers/tinkerbell/hardware/catalogue_hardware.go:54]: Use colon's to separate context from original error. The disk variable just avoids unnecessary hashing in the map which, while negligible, has a cost and is easily avoided.

```suggestion
    disk, ok := d.Disks[key]
	if !ok {
		return "", fmt.Errorf("hardware selector does not match with any given hardware machines: %v", hardwareSelector)
	}
	return disk, nil
```

**Review by chrisdoherty4 — CHANGES_REQUESTED** [review]
Great work - just some loose ends to tidy up.

File: pkg/api/v1alpha1/tinkerbelltemplateconfig_defaults.go (hunk 3 above)
Comment (panktishah26) [reply pkg/api/v1alpha1/tinkerbelltemplateconfig_defaults.go]: Yes, I am going to create a separate PR for this.

File: pkg/providers/tinkerbell/config/template-cp.yaml (hunk 5 above)
Comment (panktishah26) [reply pkg/providers/tinkerbell/config/template-cp.yaml:186]: I have added a basic validation where HardwareSelector is mandatory for each machine config for Tinkerbell.

File: pkg/api/v1alpha1/tinkerbelltemplateconfig_defaults.go [hunk 13]
```diff
@@ -29,39 +31,39 @@ warnings:
...
-func withNetplanAction(b v1alpha1.VersionsBundle) ActionOpt {
+func withNetplanAction(b v1alpha1.VersionsBundle, disk string) ActionOpt {
 	return func(a *[]tinkerbell.Action) {
 		*a = append(*a, tinkerbell.Action{
 			Name:    "write-netplan",
 			Image:   b.Tinkerbell.TinkerbellStack.Actions.WriteFile.URI,
 			Timeout: 90,
 			Environment: map[string]string{
-				"DEST_DISK": "/dev/sda2",
+				"DEST_DISK": fmt.Sprintf("%s2", disk),
```
Comment (micahhausler) [inline pkg/api/v1alpha1/tinkerbelltemplateconfig_defaults.go:66]: We don't know this will be the second disk 100% of the time

Comment (abhinavmpandey08) [reply pkg/api/v1alpha1/tinkerbelltemplateconfig_defaults.go:66]: This is the 2nd disk partition not the 2nd disk. The ubuntu image we build has UEFI is the primary partition and the actual os on the 2nd partition. That's why we are appending 2 to the disk name

File: pkg/providers/tinkerbell/create.go [hunk 14]
```diff
@@ -117,7 +117,10 @@ func (p *Provider) SetupAndValidateCreateCluster(ctx context.Context, clusterSpe
 		validator.Register(NewIPNotInUseAssertion(p.netClient))
 	}
 
-	writer := hardware.NewMachineCatalogueWriter(p.catalogue)
+	catalogueWriter := hardware.NewMachineCatalogueWriter(p.catalogue)
+
+	writer := hardware.MultiMachineWriter(catalogueWriter, p.templateBuilder.diskExtractor)
```
Comment (chrisdoherty4) [inline pkg/providers/tinkerbell/create.go:122]: This should reference the extractor from `p` not the template builder.

**Review by chrisdoherty4 — APPROVED** [review]

**Review by abhinavmpandey08 — APPROVED** [review]
/lgtm
